src/main/java/net/fabricmc/loader/metadata/ModMetadataParser.java

```diff
@@ -16,66 +16,117 @@
+	// For all intensive purposes of replicating the logic of Gson before we have migrated to Nanojson, duplicate keys will replace previous entries.
+	public static LoaderModMetadata parseMetadata(Path modJson) throws IOException, ParseMetadataException {
+		try {
+			// So some context:
+			// Per the json specification, ordering of fields is not typically enforced.
+			// Furthermore we cannot guarantee the `schemaVersion` is the first field in every `fabric.mod.json`
+			//
+			// To work around this, we do the following:
+			// Try to read first field
+			// If the first field is the schemaVersion, read the file normally.
+			//
+			// If the first field is not the schema version, fallback to a more exhaustive check.
+			// Read the rest of the file, looking for the `schemaVersion` field.
+			// If we find the field, cache the value
+			// If there happens to be another `schemaVersion` that has a differing value, then fail.
+			// At the end, if we find no `schemaVersion` then assume the `schemaVersion` is 0
+			// Re-read the JSON file.
+			int schemaVersion = 0;
+
+			try (InputStream stream = Files.newInputStream(modJson)) {
+				try (JsonReader reader = new JsonReader(new InputStreamReader(stream, StandardCharsets.UTF_8))) {
+					if (reader.peek() != JsonToken.BEGIN_OBJECT) {
+						throw new ParseMetadataException("Root of \"fabric.mod.json\" must be an object", reader);
+					}
+
+					reader.beginObject();
+
+					boolean firstField = true;
+
+					while (reader.hasNext()) {
+						// Try to read the schemaVersion
+						if (reader.nextName().equals("schemaVersion")) {
+							if (reader.peek() != JsonToken.NUMBER) {
+								throw new ParseMetadataException("\"schemaVersion\" must be a number.", reader);
+							}
+
+							schemaVersion = reader.nextInt();
+
+							if (firstField) {
+								// Finish reading the metadata
+								return ModMetadataParser.readModMetadata(reader, schemaVersion);
+							} else {
+								// schemaVersion found, but after some content -> start over to parse all data with the detected version
+								if (reader.hasNext()) {
+									while (reader.hasNext()) {
+										reader.skipValue();
+									}
+								}
+
+								break;
+							}
+						} else {
+							reader.skipValue();
+						}
+
+						firstField = false;
+					}
+
+					reader.endObject();
+				}
 			}
+
+			// Slow path, schema version wasn't specified early enough, re-read with detected/inferred version
+			try (InputStream stream = Files.newInputStream(modJson)) {
+				try (JsonReader reader = new JsonReader(new InputStreamReader(stream, StandardCharsets.UTF_8))) {
+					if (reader.peek() != JsonToken.BEGIN_OBJECT) {
+						throw new ParseMetadataException("Root of \"fabric.mod.json\" must be an object", reader);
+					}
+
+					reader.beginObject();
+					final LoaderModMetadata ret = ModMetadataParser.readModMetadata(reader, schemaVersion);
+
```

Comment:
reader.endObjec() is missing here

Reply:
It was at the end of the specific version parsers but I moved it to there